### DISCUSSION TITLE
Apollo supported deep learning models!

### DISCUSSION THREAD

**Discussion by daohu527** [ApolloAuto/apollo#15212] [Ideas]
## perception models
Starting from apollo8.0, we have **removed the model files from the source code** to facilitate quick download of the code. You should **manually download and install** the model by [Model managent](#amodel)

#### model download urls
| Model                 | Model type               | Model download address                                                                  |
|-----------------------|--------------------------|-----------------------------------------------------------------------------------------|
| petrv1_paddle        | BEV_detection            | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/petrv1.zip                      |
| horizontal_torch      | Trafficlight_recognition | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/horizontal_torch.zip            |
| quadrate_torch        | Trafficlight_recognition | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/quadrate_torch.zip              |
| vertical_torch        | Trafficlight_recognition | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/vertical_torch.zip              |
| horizontal_caffe      | Trafficlight_recognition | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/horizontal_caffe.zip            |
| quadrate_caffe        | Trafficlight_recognition | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/quadrate_caffe.zip              |
| vertical_caffe        | Trafficlight_recognition | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/vertical_caffe.zip              |
| tl_detection_caffe    | Trafficlight_detection   | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/tl_detection_caffe.zip          |
| denseline_caffe       | lane_detection           | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/denseline_caffe.zip             |
| darkSCNN_caffe        | lane_detection           | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/darkSCNN_caffe.zip              |
| cnnseg16_caffe        | lidar_3d_segmentation    | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/cnnseg16_caffe.zip              |
| cnnseg64_caffe        | lidar_3d_segmentation    | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/cnnseg64_caffe.zip              |
| cnnseg128_caffe       | lidar_3d_segmentation    | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/cnnseg128_caffe.zip             |
| center_point_paddle   | lidar_3d_detection       | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/center_point_paddle.zip         |
| mask_pillars_torch_v2    | lidar_3d_detection       | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/mask_pillars_torch_v2.zip       |
| point_pillars_torch_v2   | lidar_3d_detection       | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/point_pillars_torch_v2.zip      |
| yolox3d_onnx          | camera_detection         | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/yolox3d_onnx.zip                |
| caddn_paddle          | camera_detection         | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/caddn_paddle.zip                |
| 3d-r4-half_caffe      | camera_detection         | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/3d-r4-half_caffe.zip            |
| point_pillars_radar4d_torch | radar_detection          | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/point_pillars_radar4d_torch.zip |
| smoke_torch           | camera_detection         | https://apollo-pkg-beta.cdn.bcebos.com/perception_model/smoke_torch.zip                 |

<span id="amodel"></span>
## Model managent
We use `amodel` to manage the models in source code mode.

#### Install amodel
Since amodel 1.0 and 2.0 are not compatible, we will install it into the container later. The temporary solution now is to install it manually.
```shell
sudo pip3 install modules/tools/amodel
```

#### 1. Download model
Download the model and install it so you can use `amodel install "model_name"`
```shell
// example
amodel install center_point_paddle
```

#### 2. List models
View models already installed in the system
```shell
amodel list
```

**Comment by CesarLiu** [comment]
will apollo provide the pre-trained onnx models of yolo3d?

**Reply by daohu527** [reply]
Updated! Already add yolo3d model

Yes, will publish later on December 18th

**Comment by shy-ruqi** [comment]
```shell
[shy@in-dev-docker:/apollo_workspace]$ amodel install petrv1
WARNING:root:File downloaded before! use cached file in /tmp/petrv1.zip
Traceback (most recent call last):
  File "/usr/lib/python3.10/shutil.py", line 816, in move
    os.rename(src, real_dst)
OSError: [Errno 18] Invalid cross-device link: '/tmp/amodel/extract_path/petrv1' -> '/apollo/modules/perception/data/models/PETR_V1_paddle'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/amodel", line 8, in <module>
    sys.exit(main())
  File "/home/shy/.local/lib/python3.10/site-packages/amodel/main.py", line 52, in main
    amodel_install(args.model_name, args.skip)
  File "/home/shy/.local/lib/python3.10/site-packages/amodel/model_manage.py", line 304, in amodel_install
    is_success = _install_model(model_meta, extract_path, skip_install)
  File "/home/shy/.local/lib/python3.10/site-packages/amodel/model_manage.py", line 256, in _install_model
    shutil.move(extract_path, install_path)
  File "/usr/lib/python3.10/shutil.py", line 832, in move
    copytree(src, real_dst, copy_function=copy_function,
  File "/usr/lib/python3.10/shutil.py", line 559, in copytree
    return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
  File "/usr/lib/python3.10/shutil.py", line 457, in _copytree
    os.makedirs(dst, exist_ok=dirs_exist_ok)
  File "/usr/lib/python3.10/os.py", line 225, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/apollo/modules/perception/data/models/PETR_V1_paddle'
```



why PermissionError, how to fix?

**Reply by shy-ruqi** [reply]
I fix it by 'sudo amodel install', but when I use list info it does not need 'sudo', could explain, thx?

**Reply by daohu527** [reply]
Which Apollo version? It's a long story, I can tell you my design philosophy and why it is in the current state

**Reply by shy-ruqi** [reply]
> Which Apollo version? It's a long story, I can tell you my design philosophy and why it is in the current state

version: apollo-env-gpu:9.0-u22-preview

At the beginning, I used the following command to install:
`sudo python modules/tools/amodel/setup.py install`
successed and it works for 'amodel list', but failed for 'amodel install' and 'sudo amodel install'. Then, I rm build and use this page's way to install, it works now but need sudo. 
In my view, python pkg can install in no-root folder, just user's folder could run.
THX.

**Comment by jezzzzza** [comment]


**Reply by daohu527** [reply]
It can be partially commercialized because it uses real data. But if you want good results (usually measured by precision and recall), my suggestion is to accumulate data and train it yourself.

**Reply by daohu527** [reply]
In short, all models are based on real road test data.

**Reply by jezzzzza** [reply]
Do you also have a list of the training code for the models, to reproduce the models? Would be much appreciated.
